File: leap-year/LeapYearFSharp/Year.fs
namespace LeapYearFSharp

module Year =
    let DivisibleBy number divisor =
        number % divisor = 0
    let NotDivisibleBy number divisor =
        number % divisor <> 0
        
    let IsLeapYear year =
        match year with
            | x when DivisibleBy year 100 
                && NotDivisibleBy year 400 -> false
            | x when DivisibleBy year 4 -> true
            | _ -> false

File: leap-year/LeapYear/Year.cs
namespace LeapYear
{
    public class Year
    {
        private readonly int _year;

        public Year(int year)
        {
            _year = year;
        }

        public bool IsLeapYear()
        {
            return DivisibleBy(4) && !(DivisibleBy(100) && NotDivisibleBy(400));
        }

        private bool NotDivisibleBy(int divisor)
        {
            return _year % divisor != 0;
        }

        private bool DivisibleBy(int divisor)
        {
            return _year % divisor == 0;
        }
    }
}
